### PR TITLE
Use conditional imports for package:flutter/foundation

### PR DESCRIPTION
**Pull request by jacob314**
## Description

*Replace this paragraph with a description of what this PR is doing. If you're modifying existing behavior, describe the existing behavior, how this PR is changing it, and what motivated the change. If you're changing visual properties, consider including before/after screenshots (and runnable code snippets to reproduce them).*

## Related Issues

*Replace this paragraph with a list of issues related to this PR from our [issue database]. Indicate, which of these issues are resolved or fixed by this PR.*

## Tests

I added the following tests:

*Replace this with a list of the tests that you added as part of this PR. A change in behaviour with no test covering it
will likely get reverted accidentally sooner or later. PRs must include tests for all changed/updated/fixed behaviors. See [Test Coverage].*

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I signed the [CLA].
- [ ] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] All existing and new tests are passing.
- [ ] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [ ] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (Please read [Handling breaking changes]). *Replace this with a link to the e-mail where you asked for input on this proposed change.*
- [ ] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

File: packages/flutter/lib/src/services/font_loader.dart [hunk 1]
```diff
@@ -4,7 +4,7 @@
 
 import 'dart:async';
 import 'dart:typed_data';
-import 'dart:ui';
+import 'dart:ui' hide VoidCallback;
```
Comment (Hixie) [inline packages/flutter/lib/src/services/font_loader.dart]: why are lines like this needed? I would have expected any case where dart:ui is imported, the only other imported VoidCallback would be the same one?

Comment (jacob314) [reply packages/flutter/lib/src/services/font_loader.dart]: I've removed all these cases. They were added as part of trying to debug the cause of the strange analysis errors. As you say, there is no need to ever hide `VoidCallback` when `dart:ui` is imported.

File: packages/flutter/lib/foundation.dart [hunk 2]
```diff
@@ -39,9 +39,11 @@ export 'src/foundation/change_notifier.dart';
 export 'src/foundation/collections.dart';
 export 'src/foundation/consolidate_response.dart';
 export 'src/foundation/constants.dart';
+export 'src/foundation/core_stub.dart'
```
Comment (jakemac53) [inline packages/flutter/lib/foundation.dart]: Per offline discussion I would suggest changing this to `export 'src/foundation/core.dart';`, and renaming all these platform specific implementations so the filename starts with an `_`. The contents of the `src/foundation/core.dart` file would then just export those, like this:

```
export '_core_stub.dart'
  if (dart.library.html) '_core_web.dart'
  if (dart.library.io) '_core_io.dart';
```

File: packages/flutter/lib/src/foundation/binding.dart [hunk 3]
```diff
@@ -14,6 +14,9 @@ import 'package:meta/meta.dart';
 import 'assertions.dart';
 import 'basic_types.dart';
 import 'constants.dart';
+import 'core_stub.dart'
```
Comment (jakemac53) [inline packages/flutter/lib/src/foundation/binding.dart]: Per my other comment, this would also change to just `import 'core.dart';`

Comment (jacob314) [reply packages/flutter/lib/src/foundation/binding.dart]: +1 on @jakemac53's suggestions on this naming convention for conditional imports.

File: packages/flutter/lib/foundation.dart (hunk 2 above)
Comment (jonahwilliams) [reply packages/flutter/lib/foundation.dart]: I like this approach

File: packages/flutter/lib/src/painting/basic_types.dart [hunk 4]
```diff
@@ -41,10 +41,11 @@ export 'dart:ui' show
   TextPosition,
   TileMode,
   VertexMode,
-  VoidCallback,
   hashValues,
   hashList;
 
+export '../foundation/basic_types.dart';
```
Comment (jonahwilliams) [inline packages/flutter/lib/src/painting/basic_types.dart]: This needs to be `export `package:flutter/foundation' show VoidCallback`

File: packages/flutter/lib/src/animation/animation.dart [hunk 5]
```diff
@@ -2,7 +2,6 @@
 // Use of this source code is governed by a BSD-style license that can be
 // found in the LICENSE file.
 
-import 'dart:ui' show VoidCallback;
```
Comment (Hixie) [inline packages/flutter/lib/src/animation/animation.dart]: nit: this leaves too many blank lines

File: packages/flutter/lib/src/foundation/basic_types.dart [hunk 6]
```diff
@@ -7,8 +7,8 @@ import 'dart:collection';
 
 // COMMON SIGNATURES
 
-export 'dart:ui' show VoidCallback;
-export 'bitfield.dart' if (dart.library.html) 'bitfield_unsupported.dart';
+export 'ui_stub.dart'
+  if (dart.library.ui) 'dart:ui' show VoidCallback;
```
Comment (Hixie) [inline packages/flutter/lib/src/foundation/basic_types.dart]: it's weird that we only limit it to `show VoidCallback` on the second line... is that conditional import syntax, or should we have the `show` on both lines?

**Comment by Hixie** [comment]
We should document what we're doing here, so that people understand if they can import `dart:ui`, were they can do so, and so on.

**Comment by goderbauer** [comment]
Can you also please fill out the PR description template to document what this change does for future archeologists?

**Comment by jacob314** [comment]
@jonahwilliams can I close this PR. This was sent just as a proof of concept that the conditional imports could work. I'm on baby leave so I'll be a bit slow to respond to e-mail

**Comment by jonahwilliams** [comment]
Ahh no problem, I'll take it from here

**Comment by jacob314** [comment]
Closing as @jonahwilliams  will patch and land as this is his CL with just a few tweaks of mine on top of it.

File: packages/flutter/lib/src/foundation/basic_types.dart (hunk 6 above)
Comment (jonahwilliams) [reply packages/flutter/lib/src/foundation/basic_types.dart]: This is the correct syntax, the VoidCallback is shown from both libraries.

**Pull request closed without merging**